src/iterators/idlist_iterator.c

```diff
@@ -125,4 +119,81 @@ QueryIterator *IT_V2(NewIdListIterator) (t_docId *ids, t_offset num, double weig
   ret->Rewind = IL_Rewind;
 
   return ret;
-}
\ No newline at end of file
+}
+
+static void SetYield(QueryIterator *mr) {
+  ResultMetrics_Reset(mr->current);
+  ResultMetrics_Add(mr->current, NULL, RS_NumVal(mr->current->num.value));
+}
+
+static IteratorStatus MR_Read(QueryIterator *base) {
+  MetricIterator *mr = (MetricIterator *)base;
+  IdListIterator *it = &mr->base;
+  IteratorStatus rc = IL_Read(base);
+  if (ITERATOR_OK == rc) {
+    base->current->num.value = mr->metricList[it->offset - 1];
+    SetYield(base);
+  }
+  return rc;
+}
+
+static IteratorStatus MR_SkipTo(QueryIterator *base, t_docId docId) {
+  MetricIterator *mr = (MetricIterator *)base;
+  IdListIterator *it = &mr->base;
+  int rc = IL_SkipTo(base, docId);
+  if (ITERATOR_OK == rc || ITERATOR_NOTFOUND == rc) {
+    base->current->num.value = mr->metricList[it->offset - 1];
+    SetYield(base);
+  }
+  return rc;
+}
+
+static void MR_Free(QueryIterator *self) {
+  MetricIterator *mi = (MetricIterator *)self;
+  if (mi == NULL) {
+    return;
+  }
+  IdListIterator *it = &mi->base;
+  QueryIterator *base = &it->base;
+
+  IndexResult_Free(base->current);
+
+  if (it->docIds) {
+    rm_free(it->docIds);
+  }
+
+  if (mi->metricList) {
+    rm_free(mi->metricList);
+  }
+
+  rm_free(mi);
+}
+
+QueryIterator *IT_V2(NewMetricIterator)(t_docId *docIds, double *metric_list, size_t num_results, Metric metric_type, bool yields_metric) {
+  MetricIterator *mi = rm_new(MetricIterator);
+  IdListIterator *it = &mi->base;
+  QueryIterator *ret = &it->base;
+  mi->type = metric_type;
+  mi->metricList = metric_list;
+  it->docIds = docIds;
+  it->size = num_results;
+  it->offset = 0;
+
+  ret->lastDocId = 0;
+  setEof(ret, false);
+  ret->type = METRIC_ITERATOR;
+  ret->current = NewMetricResult();
+  // If we interested in yielding score
+  if (yields_metric) {
```

Comment:
This option is redundant.
In my opinion, we should have only one of the following options:
1. Have a single iterator type that gets a metric array on construction, and then choose between `MR_Read` and `IL_Read` (and so on) based on whether a valid array was passed or `NULL`
2. Have two explicit types (as you implemented), but the metrics iterator does not support a missing metrics array (the caller should use the id-list iterator instead)

Reply:
I changed the code such that if yields metric is not returned, then IDListIterator is returned